src/imagej/convert.py

```diff
@@ -253,6 +264,51 @@ def supports_java_to_xarray(ij: "jc.ImageJ", obj) -> bool:
         return False
 
 
+#############################
+# Java array -> NumPy array #
+#############################
+
+
+# fmt: off
+_jarraytype_map = {
+    JBoolean:     np.bool8,
+    JByte:        np.int8,
+    # JChar:      np.???,
+    JDouble:      np.float64,
+    JFloat:       np.float32,
+    JInt:         np.int32,
+    JLong:        np.int64,
+    JShort:       np.int16,
+}
+# fmt: on
+
+
+def jarray_to_ndarray(jarr):
+    """
+    Convert the given Java primitive array into a NumPy ndarray.
+
+    :param jarr: The Java primitive array
+    :return: The converted NumPy ndarray
+    """
+    assert supports_jarray_to_ndarray(jarr)
+    element_type = _jarray_element_type(jarr)
+    dtype = _jarraytype_map[element_type]
+    ndarray = np.frombuffer(memoryview(jarr), dtype=dtype)
+    return ndarray.reshape(_jarray_shape(jarr))
+
+
+def supports_jarray_to_ndarray(obj):
+    """
+    Return True iff the given object is convertible to a NumPy ndarray
+    via the jarray_to_ndarray function.
+
+    :param obj: The object to check for convertibility
+    :return: True iff conversion to a NumPy ndarray is possible
+    """
+    element_type = _jarray_element_type(obj)
+    return element_type is not None and element_type in _jarraytype_map
```

Comment:
This is super slick, but do you think it should be the default?

Is there another situation where the default is to return something that is read-only? I can't think of another If not, I'd think the principle of least surprise would suggest this should only be used if indicated by the user. Maybe with one of @elevans's keyword arguments?

Reply:
That is a good point! The fact that it is read-only is unfortunate. However, the old default is prohibitively slow for arrays of any significant size. So what's better default behavior: that the converted primitive array is surprisingly read-only and not a Python list? Or that the result is a read-write Python list but took minutes to complete the conversion?

A third option would be another Python-side *view* on the Java array—i.e. don't copy anything, but only provide a Python class as a window into the primitive array data. So the conversion call itself would be fast, but then if you actually iterated the entire converted collection it would be glacially slow again to complete the loop, as per the current default.

Reply:
> A third option would be another Python-side view on the Java array—i.e. don't copy anything, but only provide a Python class as a window into the primitive array data. So the conversion call itself would be fast, but then if you actually iterated the entire converted collection it would be glacially slow again to complete the loop, as per the current default.

I actually kind of like this idea. What if we made a view object whose `__setitem__` fills in some data structure `overwritten`, and `__getitem__` reads from the memory view if `overwritten` doesn't have an element at that position? Does that make sense?

Reply:
> Does that make sense?

Yeah, interesting idea. Then we would retain fast read access to the original data (although there would be a hash test at every element access now), while having write access also. This feels like something more general, though... wrapping a read-only numpy array into some kind of copy-on-write object or similar. I poked around a little, but didn't find anything definitive in this space.

A *fourth* option would be to wrap the numpy array in the readonly view, then make a copy with the copy-slice syntax `[:]`.

Thinking about it more: I am in favor of KISS here and just returning the read-only data, and if it bites someone, they can copy-slice it themselves. This is the most performant and least magical option.

Reply:
> Thinking about it more: I am in favor of KISS here and just returning the read-only data, and if it bites someone, they can copy-slice it themselves. This is the most performant and least magical option.

Sure. But is it the least surprising?

I'm all in favor of waiting on this idea until there is demand for it. But my two cents is:
* AFAIK pyimagej's strategy is to a return new object *or* a R/W view.
* This behavior would introduce a new default behavior, a read-only view.

I'll delegate to your judgement, but here my lingering questions:

* What's the easiest way to replicate this behavior in a read/write way? If it's something we could expect the average user to do (or something that the average user wouldn't bother calling `ij.py.to_python` to do), then let's leave it as is.

* The other question: when is a user likely to use this? What's the use case? Are they gonna want to pass this to some mutating function, or to some sort of visualization tool?

Reply:
> Sure. But is it the least surprising?

PoLA and KISS principles are in conflict here. Probably the least astonishing thing would be to invent a `JavaArray` wrapper backed by the Java primitive array, which is a read/write view. This would behave analogously to the Java collection wrappers that scyjava provides. I like that option for its consistency with the other solutions.

But what I don't like is that it wouldn't expose this fast read-only option in a discoverable way. Using a kwargs hint would work, but it means *no one will ever use this functionality by default*. They will need to read the docs or a tutorial in order to discover it. They are likely to instead just call `scyjava.to_python` on their array naively, get horrible performance when iterating the result, and then bang their head around for a few hours trying to figure out a better way via SO searches and so forth. It's not a good situation. Whereas if we make the fast option the default, people will get that behavior immediately and be happy it's so fast. And then when they try to write to it, get an error about how it's read-only, and they are likely to think to make a copy in that case. So while I'm not sure which behavior is the least *astonishing*, I think the read-only wrapper will be the least *frustrating*.

Reply:
Sounds good, thanks for thinking about these things!

Reply:
This is a really interesting discussion and I think you both make very good points. Putting on my scientist hat, I think I would prefer a faster compute time with a _read-only_ output over a slower computer time with a _read-write_ object. This would allow me to decide when and where to do the expensive copy operation (if my workflow needs to do so). And I agree if this function was not default most of our users will likely not use it and gain the performance boost the `memoryview` supplies.

This naive question, but can we convert the Java arrays to a `bytearray` or whatever Java calls it's buffer protocol? My understanding is that Python's `memoryview` with a buffer allows read/write. This works fine:

```python
>>> barr = bytearray([5, 10, 3])
>>> print(barr)
bytearray(b'\x05\n\x03')
>>> view = memoryview(barr)
>>> print(view[1])
10
>>> print(barr[1])
10
>>> view[1] = 15
>>> print(view[1])
15
>>> print(barr[1])
15
```
The documentation is still a little above my head, but it seems that `memoryview` just needs an writable buffer.